From 2b9ee60d58f497c43ee9c86a17602e1c3ffc8530 Mon Sep 17 00:00:00 2001
From: Cheng Jian <cj.chengjian@huawei.com>
Date: Sat, 29 May 2021 14:42:47 +0800
Subject: [PATCH] livepatch/core: Add livepatch consistency depends
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: 2b9ee60d58f497c43ee9c86a17602e1c3ffc8530


euler inclusion
category: feature
Bugzilla: 51921
CVE: N/A

----------------------------------------

Now, arm64 don't support DYNAMIC_FTRACE_WITH_REGS and RELIABLE_STACKTRACE.
which the first is necessary to implement livepatch with ftrace and the
second allow to implement per-task consistency.

So. arm64 only support LIVEPATCH_WO_FTRACE and STOP_MACHINE_CONSISTENCY.
but other architectures can work under LIVEPATCH_FTRACE with
PER_TASK_CONSISTENCY. commit the depends to avoid incorrect configuration.

Signed-off-by: Cheng Jian <cj.chengjian@huawei.com>
Reviewed-by: Li Bin <huawei.libin@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>

Signed-off-by: Dong Kai <dongkai11@huawei.com>

Signed-off-by: Ye Weihua <yeweihua4@huawei.com>
Reviewed-by: Yang Jihong <yangjihong1@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 kernel/livepatch/Kconfig | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/livepatch/Kconfig b/kernel/livepatch/Kconfig
index 354fdfdebc32..297ca41c695e 100644
--- a/kernel/livepatch/Kconfig
+++ b/kernel/livepatch/Kconfig
@@ -39,12 +39,14 @@ config LIVEPATCH_FTRACE
 	bool "based on ftrace"
 	depends on HAVE_LIVEPATCH_FTRACE
 	depends on DYNAMIC_FTRACE_WITH_REGS
+	select LIVEPATCH_PER_TASK_CONSISTENCY
 	help
 	  Supports kernel live patching based on ftrace
 
 config LIVEPATCH_WO_FTRACE
 	bool "without ftrace"
 	depends on HAVE_LIVEPATCH_WO_FTRACE
+	select LIVEPATCH_STOP_MACHINE_CONSISTENCY
 	help
 	  Supports kernel live patching without ftrace
 
@@ -60,6 +62,7 @@ choice
 
 config LIVEPATCH_PER_TASK_CONSISTENCY
 	bool "per task consistency"
+	depends on LIVEPATCH_FTRACE
 	help
 	  Use basic per-task consistency model
 	  It's a hybrid of kGraft and kpatch:
@@ -70,6 +73,7 @@ config LIVEPATCH_PER_TASK_CONSISTENCY
 
 config LIVEPATCH_STOP_MACHINE_CONSISTENCY
 	bool "stop machine consistency"
+	depends on LIVEPATCH_WO_FTRACE
 	help
 	  Use stop machine consistency model
 	  stop-machine consistency and kpatch's stack
-- 
2.26.2

